hail/src/main/resources/include/hail/Decoder.h

```diff
@@ -209,6 +211,50 @@ class LEB128InputBuffer {
     bool read_boolean() { return buf_.read_boolean(); }
 };
 
+template<typename Decoder>
+class ReadIterator {
+private:
+  std::shared_ptr<Decoder> dec_;
+  Region * region_;
+  NativeStatus * st_;
+  char * value_;
+
+public:
+  ReadIterator() :
+  dec_(nullptr),
+  region_(nullptr),
+  st_(nullptr),
+  value_(nullptr) { }
+
+  ReadIterator(std::shared_ptr<Decoder> dec, Region * region, NativeStatus* st) :
+  dec_(dec),
+  region_(region),
+  st_(st),
+  value_(nullptr) { }
+
+  ReadIterator<Decoder>& operator++() {
+    if (dec_ != nullptr) {
+      if (dec_->decode_byte(st_)) {
+        value_ = dec_->decode_row(st_, region_);
+      } else {
+        value_ = nullptr;
+        dec_ = nullptr;
+      }
+    }
+    return *this;
+  }
+
+  bool operator==(ReadIterator other) const {
+    return (dec_ == other.dec_) && (value_ == other.value_);
+  }
+
+  bool operator!=(ReadIterator other) const { return !(*this == other); }
+
+  char * operator*() { return value_; }
```

Comment:
This should probably be a const iterator, right? Also, this should return a reference, otherwise to get the actual value you'd have to write `**it`. So, this should return `char const&` (or `const char&` if you prefer).

Reply:
the `char *` is the "value, though, right? We don't actually want to dereference the returned pointer, since in most cases that makes no sense. Unless you're talking about something else?

Reply:
Oh of course, never mind! Then it should be `char const*`, a non-const pointer to const data. And eventually, once we have more C++ infrastructure, this might return an object that represents a hail type wrapping the pointer, not just a raw pointer.

Reply:
ok, sounds good!